pilot/pkg/serviceregistry/kube/controller/controller.go

```diff
@@ -439,9 +446,51 @@ func (c *Controller) HasSynced() bool {
 		!c.nodeInformer.HasSynced() {
 		return false
 	}
+
+	// after informer caches sync the first time, process resources in order
+	// TODO(landow) is it possible to start the informers one-by-one in order?
+	c.syncMu.Lock()
+	defer c.syncMu.Unlock()
+	if !c.initialSync {
+		if err := c.SyncAll(); err != nil {
+			log.Errorf("one or more errors force-syncing resources: %v", err)
+		}
+		c.initialSync = true
+	}
+
 	return true
 }
 
+func (c *Controller) SyncAll() error {
```

Comment:
One concern: say we get the updates for some service foo, v1->v2->v3.

Today we process these v1, then v2, then v3.

Now we will process in some nondeterminstic ordering: v1,v2,v3,v1,v2,v3 is possible, or interleaved v1,v2,v1,v3,v2,v3 or whatever.

Specifically concerning is where we undo things, going from v2->v1

Given we are not marked ready until all of these are complete, I think this should be fine though? since the last event will always be v3? Note my example was looking at a single object, this of course will have multiple objects in reality but they all follow the same rule that we always process the last object last.

K8s gets around this by putting the key in the queue, not the object, and the event process the latest version. So we may just process v3,v3,v3,v3,v3,v3 (or v1,v2,v3,v3,v3,v3, depending on how fast things run). This would be a big change for us though.

Reply:
Using [workqueue](https://godoc.org/k8s.io/client-go/util/workqueue) could help with that.

One other approach to this whole thing would be:
1. OnEvent, check if informer caches have synced. If they haven't yet, put the event onto a separate priority queue. 
2. Once caches sync, start processing that priority queue which should guarantee correct ordering. Any new events will end up in the regular queue. 
3. When priority/init queue is drained, start processing things normally.

Reply:
workqueue would only help if we add things to the queue faster than we process them, right? it actually may break the ordering by the deduping I *think* without digging into the implementation. Basically if we enter v2,v1,v2 into the queue, it would dedupe to v2,v1 which is backwards order?

A similar idea is we could actually put a "signal" onto the queue. Once informer.HasSynced is true, we input a SyncSignal into the queue. In the processing, if we see a SyncSignal instead of Service or whatever, we will mark some field. Once all signals have been processed, we know everything has been processed. I guess that doesn't solve the out of order services/endpoints though.

Reply:
Hm I think we can fix the ordering issue by checking the informers cache at the time of processing. We'd end up with some duplicated work without adding workqueue, but that would take some effort to add.